### DISCUSSION TITLE
supabaseUrl or supabaseKey required

### DISCUSSION THREAD

**Discussion by CaseyG5** [supabase/supabase#3218] [Questions] [answered]
1. `Error: supabaseUrl is required.`
2. `Error: supabaseKey is required.`

SupabaseClient.ts:
`   51 | this.storageUrl = `${supabaseUrl}/storage/v1`;`
`> 52 | this.schema = settings.schema;`

These two errors I'm getting in my React project (one at a time, not both together) are rather frustrating since my supabaseUrl and supabaseKey are, and have always been, provided to the `createClient()` function or `SupabaseClient()` constructor.  For the first error, Supabase complains about line 51 in it's own file.  For the second, line 52 is a problem.

I updated `@supabase/supabase-js` from version 1.22.5 to 1.22.6 and the URL error disappeared temporarily but then I switched to using my Service key instead of my Anon key and now the KEY error has taken its place.  Is this a Supabase bug that is plaguing anyone else?  Thanks for any tips you might have.

Casey

**Comment by dshukertjr** [comment, marked as answer]
Hi @CaseyG5!

I am sorry that you are having some trouble. I just upgraded one of my Next.js project to use both v1.22.5 and v1.22.6 of `supabase-js`, and was not able to reproduce this issue with both annon key and service key. Do you know if this issue happened on earlier versions of `supabase-js`?

Could you share your code where you are calling `createClient()`?

**Reply by CaseyG5** [reply]
Hey, thanks for your reply, dshukertjr.  Not sure about previous versions, but here's what's in my supabaseClient.js file where I call `createClient()`...

`import { createClient } from '@supabase/supabase-js';`

`const supabaseUrl = process.env.REACT_APP_SUPABASE_URL;`
`const supabaseAnonKey = process.env.REACT_APP_SUPABASE_ANON_KEY;`
`const supabaseServKey = process.env.REACT_APP_SUPABASE_SERV_KEY;`

`const supabase = createClient(supabaseUrl, supabaseServKey);`

`export default supabase;`

I'm not getting the error anymore (knock on wood).  But the only change I made was to name my `.env` variables so they begin with `REACT_APP_SUPABASE_`.  I don't understand how that would make a difference but also I don't know the inner workings of `process.env`.

**Reply by dshukertjr** [reply]
@CaseyG5 

Ah, I see! That was probably the issue then! Environmental variables in React has to start with `REACT_APP_` as explained [here](https://create-react-app.dev/docs/adding-custom-environment-variables/). 
I am glad to hear that everything is working now 😄

**Reply by CaseyG5** [reply]
Yup that explains it then.  Thanks, I didn't know React was picky about that too.  Such an idiosyncratic framework.

**Reply by stephencarr** [reply]
I am having a similar problem:

```
# .env
REACT_APP_SUPABASE_URL="https://foo.supabase.co"
REACT_APP_SUPABASE_ANON_KEY="rAnDOM"
```

```
# initSupabase.js

import { createClient } from '@supabase/supabase-js';

const options = {
  schema: 'public',
  autoRefreshToken: true,
  persistSession: true,
  detectSessionInUrl: true,
};

const supabaseUrl = process.env.REACT_APP_SUPABASE_URL;
const supabaseAnonKey = process.env.REACT_APP_SUPABASE_ANON_KEY;

// Create a single supabase client for interacting with your database
export const supabase = createClient(supabaseUrl, supabaseAnonKey, options);
```

This always results in

```
Unhandled Runtime Error
Error: supabaseUrl is required.
```

If I `console.log(supabaseUrl)` I get the string I am env expecting.

**Reply by dshukertjr** [reply]
@stephencarr Hmm, that is weird... 

If you try directly enter the `supabaseUrl` and `supabaseAnonKey` into `createClient()`, would it work?

```typescript
export const supabase = createClient('https://foo.supabase.co', 'rAnDOM', options);
```

If the above code does not produce the error, we can at least say that it's a react config problem and not a Supabase problem.

**Reply by stephencarr** [reply]
Yes that works. I should mention I am using NextJS, not vanilla React.

**Reply by kk21** [reply]
For nextjs, u can put the env vars in .env.local and use the prefix "NEXT_PUBLIC_" for env vars to be exposed to the web browser.
https://nextjs.org/docs/basic-features/environment-variables

**Reply by gdickison** [reply]
I'm having the same problem using Next JS. My env variables are properly prefixed with NEXT_PUBLIC_. I have another Next JS project that I attached to Supabase and it works fine. However, I created a new project using the [Quickstart: Next JS](https://supabase.com/docs/guides/with-nextjs) tutorial and createClient() can't read the variables. If I use the URL and Key without variables it works, so I know the problem is not with the API.

**Reply by JhaAman** [reply]
Still having this issue with Next.js.

I have my env variables in a `env.local` file, all NEXT_PUBLIC_ variables. My client code looks like:

``` TS
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL as string;
const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY as string;

const supabase = createClient(supabaseUrl, supabaseAnonKey);
```

And there's always and error in the browser saying it won't work. It is working if I directly put in the url and key, but not with env variables, even though all the tutorials have that and it's working.

**Reply by JhaAman** [reply]
Figured it out - for future Next and Supabase developers, be sure to restart `yarn dev` and restart it to reload the environment variables.

**Reply by gdickison** [reply]
Restarting (`npm run dev`) did not help me, unfortunately.

**Reply by SmithJohnTaylor** [reply]
@gdickison Make sure your env file has a period in front of it `.env.local` not `env.local`
Just ran into this issue myself

**Reply by gdickison** [reply]
Thanks @SmithJohnTaylor. I do have that file named correctly. Still getting the error. I just started another project from scratch - same thing.

**Reply by gdickison** [reply]
Well, on my new project I was getting the same error. Then I went back through this thread and tried (again) what others have suggested:
- assigned the environment variables to constants before passing them to createClient
- shutting down and restarting the server by running npm run dev again

Now it all works. Go figure.

**Reply by Prudence97** [reply]
> @gdickison Make sure your env file has a period in front of it `.env.local` not `env.local`
> Just ran into this issue myself

This was the first mistake I made. Bow I'm trying to get both Anon Key and Url defined

**Reply by Bianxzy20** [reply]
what do you guys mean by shutting down the server?

**Reply by Felix-Okeyo** [reply]
like run a "ctrl + c" command in the bash/zsh terminal window the server is running at and then running "npm run dev" again to work on the just-installed dependencies

**Reply by sahils0** [reply]
I faced the same issue and indeed it was due to the .env.local variable and i just copied the env variable names and pasted it into my main code and it worked like a charm

**Reply by ajhous44** [reply]
Also have the same issue when using .env.local

**Comment by MildTomato** [comment]
Is it possible there’s a trailing slash in the URL?

**Reply by CaseyG5** [reply]
Didn't have any trailing slash.  Thanks for the thought though.

**Comment by daniellaera** [comment]
having the same issue

**Reply by sai-mamidi** [reply]
it really helped bro, Thank you!

**Comment by Ramoncarden** [comment]
Hello just wanted to chime in here. I had this same problem and realized my mistake was that I accidentally put my .env.local file in my src folder instead of project route. Just in case anyone is having this issue.

**Reply by fourwhitesocks** [reply]
@Ramoncarden 
Thank you, Made the same mistake ;-)  save me!

**Reply by torleado** [reply]
I was making this mistake too, thank you so much!

**Reply by nahanez22** [reply]
me paso exactamente igual

**Reply by oozkanneray** [reply]
oh thanks

**Reply by buntaga** [reply]
I had the exact same issue, thanks!

**Reply by andreihuyoa** [reply]
thanks!

**Reply by Felix-Okeyo** [reply]
In my case, I even restarted the supabase installation, restarted the project, got the typescript sql file again. Even redid the .env.local file as well. All along, I had my .env.local file inside of my public folder instead of root folder

**Reply by Oms-343** [reply]
> Hello just wanted to chime in here. I had this same problem and realized my mistake was that I accidentally put my .env.local file in my src folder instead of project route. Just in case anyone is having this issue.

thanks man for the clutch

**Reply by chaosprint** [reply]
lol same here

**Reply by chamals3n4** [reply]
haa haa thanks dude :)

**Comment by Qikta** [comment]
Hello.
I was able to get it to work by setting the ENV file as follows.
I have added 'NEXT_PUBLIC_'. I hope this helps someone else.

`.env.local`
```
NEXT_PUBLIC_SUPABASE_URL=*****
NEXT_PUBLIC_SUPABASE_ANON_KEY=*****
```

`utils/supabaseClient.ts`

```ts
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL as string
const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY as string
export const supabase = createClient(supabaseUrl, supabaseAnonKey)
```

**Reply by joedeluca** [reply]
helped me! thanks Qikta.

**Reply by DragutinDjokovic** [reply]
Thanks.

**Reply by richewinqred** [reply]
Thanks, worked for me.

```
import { createClient } from "@supabase/supabase-js";

const supabaseUrl = import.meta.env.VITE_REACT_APP_SUPABASE_URL as string;
const supabaseAnonKey = import.meta.env.VITE_REACT_APP_ANON_KEY as string;

export const supabase = createClient(supabaseUrl, supabaseAnonKey);

export default supabase;
```

**Reply by meshackm** [reply]
I was facing same issue, Thanks a lot.

**Reply by kimutai-alex** [reply]
when I almost gave up on my project, I thought let me try this, and it worked.

**Reply by zillBoy** [reply]
Hi, this does work but this way your `SUPABASE_URL` & `ANON_KEY` are going to be publically available.
You can check your: browser developer console > Sources > webpack_internal:// -> now we got access to the whole codebase and can see the URL. 

So, this is a security issue, I'm looking for this solutions at the moment.

**Reply by zillBoy** [reply]
FYI, I found this. What we need to do is, set up the RSL Policy properly. `SUPABASE_URL` & `ANON_KEY` can be exposed publically.
https://github.com/orgs/supabase/discussions/6877#discussioncomment-4907344

**Comment by yuxxeun** [comment]
i faced the same thing after deploying, but in my case i was using `SvelteKit` and i read some related articles that in the `.env` file for svelte using `VITE_SUPABASE_URL` and `VITE_SUPABASE_ANON_KEY` but still error

**Reply by yuxxeun** [reply]
> i faced the same thing when after deploying, but in my case i was using `SvelteKit` and i read some related articles that in the `.env` file for svelte using `VITE_SUPABASE_URL` and `VITE_SUPABASE_ANON_KEY` but still error

and this code in the `lib/supabaseClient.js` : 
```js
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = import.meta.env.VITE_SUPABASE_URL
const supabaseAnonKey = import.meta.env.VITE_SUPABASE_ANON_KEY

export const supabase = createClient(supabaseUrl, supabaseAnonKey)
```

**Reply by Gilgahex** [reply]
I have the same issue, did you find a solution?

**Reply by yuxxeun** [reply]
Yeah finally, I'm find a solution for this case.

```js title=""
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = import.meta.env.PUBLIC_SUPABASE_URL
const supabaseKey = import.meta.env.PUBLIC_SUPABASE_KEY

export const supabase = createClient(supabaseUrl, supabaseKey)
```
```env
PUBLIC_SUPABASE_URL=your_public_url
PUBLIC_SUPABASE_KEY=your_key
```

**Comment by psnwd** [comment]
Hello, IDK why but just add **prefix**

```
PUBLIC_SUPABASE_URL = the_url
PUBLIC_SUPABASE_ANON_KEY = long_key
```
to
```
NEXT_PUBLIC_SUPABASE_URL = the_url
NEXT_PUBLIC_SUPABASE_ANON_KEY = long_key
```

BlackCAT

**Reply by redpwilliams** [reply]
Working primarily with Next.js, prefixing `NEXT_PUBLIC_` to an environment variable makes it accessible to the browser rather than just the server:

> By default environment variables are only available in the Node.js environment, meaning they won't be exposed to the browser.
> 
> In order to expose a variable to the browser you have to prefix the variable with **NEXT_PUBLIC_.** For example:
> 
> ```js
> NEXT_PUBLIC_ANALYTICS_ID=abcdefghijk
> ```
> 
> This loads process.env.NEXT_PUBLIC_ANALYTICS_ID into the Node.js environment automatically, allowing you to use it anywhere in your code. The value will be inlined into JavaScript sent to the browser because of the NEXT_PUBLIC_ prefix. This inlining occurs at build time, so your various NEXT_PUBLIC_ envs need to be set when the project is built.

From https://nextjs.org/docs/basic-features/environment-variables

I imagine it is quite similar for the React case.

The only issue is security; By exposing the secrets to the browser, anyone using your site would be able to see the javascript code - specifically where you use this environment variable - and get the actual value of the secret.

**Reply by Capta1nRaj** [reply]
That's what I was looking for, so I need to ask, is it safe to expose the Supabase keys? I think their docs are not that good in terms of security.

**Reply by zillBoy** [reply]
Hi @redpwilliams, true. Where you able to solve this security issue. 
Is there a way other than using the `NEXT_PUBLIC_` prefix

**Reply by zillBoy** [reply]
FYI, I found this. What we need to do is, set up the `RSL Policy` properly. 
https://github.com/orgs/supabase/discussions/6877#discussioncomment-4907344

**Reply by laoskey** [reply]
i trid some ways to  save this problem,it works for me. Thanks bro

**Comment by landmann** [comment]
Btw - if anyone is working with Vercel, make sure to add your environment variables to the Vercel dashboard! Spent so much time trying to fix it locally...

**Reply by DavidSabine** [reply]
What names did you use? I'm using PUBLIC_SUPABASE_URL and PUBLIC_SUPABASE_ANON_KEY but my site (on Vercel) isn't getting those values at runtime.

**Reply by landmann** [reply]
Gotta do NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY 😉
Make sure to change them in your code as well...

**Comment by istir** [comment]
If you're using Vercel you might've set environmental variables only for `production` environment and are building in for example in `Preview` environment (as a branch normally would build first)

**Reply by guylepage3** [reply]
I'm assuming one should change the SUPABASE_URL and not the NEXT_PUBLIC_SUPABASE_URL. Correct?

**Reply by Capta1nRaj** [reply]
I think there are bucket's policies which makes it useless for others to use our keys.

**Comment by Bumboobee** [comment]
I was facing the same issue, in one ```React``` project. After spend like half of the day, doing some research, I just realease that **every** modify that you make in your ```.env``` file, you just restart the server. I you change one letter, restart the server, no matter how small your change was, **YOU MUST** restart you dev build. And of course, check in your  ```.env``` file is in the root of the project!

Good code to everyone!

**Comment by ivanosquis10** [comment]
Tenia el mismo problema, use vercel para el deploy, solo tuve que colocar en la configuracion del proyecto las variables de entorno del proyecto y se quita el error

I had the same problem, I used vercel for the deploy, I only had to put in the project configuration the project environment variables and the error is removed.

**Reply by yuxxeun** [reply]
can i see your `.env` files and `supabase` configuration file?
and tell me what is your project framework? because different framework have different name variabel

**Comment by kotesh-arya** [comment]
**Hello there**  👋🏻

I have encountered the same issue and resolved it by wrapping the `REACT_APP_ANON_KEY` inside the double quotes as a string and placing the `.env` file in the root level, not in the `src` folder.

![image](https://user-images.githubusercontent.com/69259490/227478477-66e2845a-bf13-487b-822d-3647a9d3e03c.png)

 🚨 **Whenever you do changes in the `.env` file, please **restart** the server, or else it will not detect the latest changes in the file.**

**Reply by Bumboobee** [reply]
is this a random key right? Please, if it is **your real api key I recomend** that you don’t share here or in any where else on the internet. If your database contains valuable information, the data can be access by other people with this api key that shows in the image! 🤙
**Be aware that your api key is a way to acess you database!!**

**Reply by MatousAc** [reply]
@Bumboobee I am considering publicly showing my API key. I only have a PUBLIC_SUPABASE_URL and PUBLIC_SUPABASE_ANON_KEY. My database does not have  *any* sensitive_ information in it nor do I think there will ever be a need to store such things for my app. I also have Supabase RLS set up so users can only ever see their own information in tables.

In my case, is it safe to publicly expose my PUBLIC_SUPABASE_URL and PUBLIC_SUPABASE_ANON_KEY? I need them to communicate with Supabase, but I am offering my app as a [package on npm](https://www.npmjs.com/package/unitplayground) and I need everything, including environment variables, to be included in the package.

**Reply by Bumboobee** [reply]
@MatousAc How are you?
In your case, **I see no problem not "hiding" your API information**. This might be a different scenario if the user can access **all** the datastore in the database. If the user is a little more programming savvy, he can access his data and simply delete **all** the data of all users within it.

But as you apply the RLS rule, this can no longer be done. The user will only be able, if he so wishes, to delete his own data, therefore, it is his responsibility...

If you are working with a page, static or not, and you make the deploy in, for example  netlify service, the offer a set **enviroment variables** option. *Thats basic hide you enviroment variables right into the deploy of your application.*

🗽 **[see this here](https://docs.netlify.com/configure-builds/environment-variables/)**

I don't know if you can do this with your package in npm, but it's worth taking a look if the service you are using to store your app, there is an option to set your environment variables, in your case, *PUBLIC_SUPABASE_URL and PUBLIC_SUPABASE_ANON_KEY*

**Reply by MatousAc** [reply]
@Bumboobee for offering your expertise. Yes, if a user really wishes to delete all their data, I don't care if they do that. I will offer an option to delete their account anyways . . .

I've never used Netlify before. I'll see if that can help. I'm writing a SvelteKit package. They say they offer ways to include .env variables in the build . . . but I can't figure out how to do it.

Thanks again.

**Reply by MatousAc** [reply]
@Bumboobee Update: So I was demonstrating how to use `CompletableFuture`s in Java for another class and I fetched the html to the front page of my app. Lo and behold, it returned:
```
... html ...
<script>
{
	__sveltekit_1bohu6o = {
		env: {"PUBLIC_SUPABASE_ANON_KEY":"my_anon_key","PUBLIC_SUPABASE_URL":"my_url"},
		base: new URL(".", location).pathname.slice(0, -1),
		element: document.currentScript.parentElement
	};

...
}
</script>
```

I followed the default SvelteKit way of handling environment variables, `$env/static/public`, documented [here](https://kit.svelte.dev/docs/modules#$env-static-public). I thought the environment variables would be hidden, but instead they're extremely easy to get.
So unless I'm seriously screwing up somewhere else, it seems that these "public" environment variables are truly public. Anyone can see them easily. So, at this point, I really don't see a need to hide my environment variables at all. I still see why they are useful for deploying in different environments, but I don't see any reason whatsoever to try to hide them.

**Reply by Bumboobee** [reply]
How's going @MatousAc? sorry for the delay to reply :/

So I enter in the link that you provider on how to set a enviroment variable... I don't know Svelt verry well, but looks like the correct way to set your secrets key is using ```$env/dynamic/private```, as the showing in this section of Svelt [documentation](https://kit.svelte.dev/docs/modules#$env-dynamic-private).

And that make sense, the prefix **public** in ```$env/static/public``` really is gonna make your secret key, not secret anymore, looks like using the **public** prefix, the client can access your secret without any trouble. 

Worth give a look at it 🤙
This **[stack overflow discussion](https://stackoverflow.com/questions/68479217/how-to-load-environment-variables-in-svelte-using-vite-or-svite)** may help too, 

**++** I don't know if you are usign Vite in your project, but look like he can make your work a lot of easier. 

> The import should look like this: 
~~~ JavaScript
import { API_KEY } from '$env/static/private';
~~~

Best regards 🤠!!

**Reply by MatousAc** [reply]
Thanks. Yeah . . . boy is this confusing. All the tutorials and directions I looked at used the public prefix, so I assumed I was supposed to do that too. Maybe that was my issue. Anyways, doing that would bring me back to other problems, so at this point I think I will just leave things public I guess.

Thanks for all the input!

**Reply by kotesh-arya** [reply]
> is this a random key right? Please, if it is **your real api key I recomend** that you don’t share here or in any where else on the internet. If your database contains valuable information, the data can be access by other people with this api key that shows in the image! 🤙 **Be aware that your api key is a way to acess you database!!**

Don't worry, i just changed few characters while sniping this shot 😜

**Comment by ghost** [comment]
When I build with next.js13.4.2, I get "Error: supabaseUrl is required." However, in the local environment, such an error does not occur and it works properly. Has anyone had a similar issue and resolved it?
.env.local exists in the root directory

(.env.local)
NEXT_PUBLIC_SUPABASE_URL= https://jhpcnxpbqfqeociocoan.supabase.co
NEXT_PUBLIC_SUPABASE_ANON_KEY= Projectkey


(supabase.ts)
import { createClient } from '@supabase/supabase-js'
const supabase = createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL as string,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY as string,
  );
export default supabase

**Reply by Bumboobee** [reply]
Hey, since it's working properly in your local environment, what service are you exactly using to deploy your production application?

Some deploy Services have a especif way to configure enviroment variables, such **[Vercel](https://vercel.com/docs/concepts/projects/environment-variables)** and **[Netlify](https://docs.netlify.com/configure-builds/environment-variables/)**

The error you are getting may be because these environment variables are not set in the service you are using to deploy you application.

**Reply by ghost** [reply]
It's vercel!

**Reply by Bumboobee** [reply]
See if you have configured the enviroments viriables correctly **[here](https://vercel.com/docs/concepts/projects/environment-variables)**.

**Reply by ghost** [reply]
thank you. i will try

**Reply by ghost** [reply]
I was able to build successfully! thank you very much!

**Comment by swyxio** [comment]
in my case the error was `Error: supabaseKey is required.`  but the cause/fix was the `SUPABASE_SERVICE_ROLE_KEY` was empty. misleading error name.

**Reply by Tholkappiar** [reply]
Hey could you help me here in this [issue](https://github.com/orgs/supabase/discussions/3218#discussioncomment-7150182) . Thanks in Advance .

**Reply by vinicius33** [reply]
Thanks for this, it helped me a lot 🙇🏼 . This error is so misleading.

**Comment by DAOUD-Nadjem-eddine** [comment]
just create a project in supabase and you will get the anon key ad url,put them in the create client
if you are in nuxt just follow the doc here with a demo project you will get what i said: https://supabase.com/docs/guides/getting-started/quickstarts/nuxtjs

**Reply by Tholkappiar** [reply]
yeah figured that out lately ! thanks !

**Comment by Felix-Okeyo** [comment]
remove the " " from your environment variables.

**Comment by Capta1nRaj** [comment]
Is it safe to expose the Supabase keys? Because NEXT_PUBLIC expose the keys in public, I think their docs are not that good in terms of security.

**Reply by Serginyo90** [reply]
> Is it safe to expose the Supabase keys? Because NEXT_PUBLIC expose the keys in public, I think their docs are not that good in terms of security.

Did you find the answer? Looks like it is not safe

**Reply by Capta1nRaj** [reply]
> > Is it safe to expose the Supabase keys? Because NEXT_PUBLIC expose the keys in public, I think their docs are not that good in terms of security.
> 
> Did you find the answer? Looks like it is not safe

Nah mate, I just shifted back to my MonogDB, it's best for me. BTW currently learning AWS DynamoDB too.

Note: Not confirmed, but there is no issue in making the key public cuz you have to set the bucket policy which will make it 200% safe. (As per my knowledge.)

**Comment by mostafahamedbesher** [comment]
this problem happened to me although i was correctly named the environment variable file(.env.local) and also correctly named the variables in this file correctly
and the problem happened because i was importing server component in a client component which is forbidden as i read that in nextjs docs
see this link:
https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#unsupported-pattern-importing-server-components-into-client-components

so when i fixed this , this works well and that error disappeared
and you can fix this in this way in this link : 
https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props

**Comment by ajbspinhope** [comment]
Very helpful comments here. Thanks for the contributions. I have got similar error and finally managed to resolved it. Here are my settings.

I am using Next.js framework.

Created .env file. This is placed under the root of your project.

NEXT_PUBLIC_SUPABASE_URL=https://yourdb.supabase.co
NEXT_PUBLIC_SUPABASE_ANON_KEY="Here place your anonymous key"
NEXT_PUBLIC_SUPABASE_SERVICE_ROLE_KEY="Here place your service role key"

Next I have created a directory called lib
1. Created a file called supabase.ts

import { createClient } from "@supabase/supabase-js";

export const supabase = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL!,
  process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!
);

export const supabaseAdmin = createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_SERVICE_ROLE_KEY!
);


2. Created a file called supabaseClient.ts

import { createClient } from '@supabase/supabase-js'

const supabaseUrl = process.env.PUBLIC_SUPABASE_URL as string;
const supabaseAnonKey  = process.env.PUBLIC_SUPABASE_ANON_KEY as string;

console.log("Supabase URL:", supabaseUrl);  // Check if this prints the correct value
console.log("Supabase AnonKey:", supabaseAnonKey );  // Check if this prints the correct value

if (!supabaseUrl || !supabaseAnonKey) {
  throw new Error("Supabase URL or Anon Key not set in environment variables");
}

const supabase = createClient(supabaseUrl, supabaseAnonKey);

export default supabase;


3. Created a file called supabaseAdmin.ts file
// lib/supabaseAdmin.js

import { createClient } from '@supabase/supabase-js'

const supabaseUrl = process.env.PUBLIC_SUPABASE_URL as string;
const supabaseServiceRoleKey = process.env.SUPABASE_SERVICE_ROLE_KEY as string;

console.log("Supabase URL:", supabaseUrl);  // Check if this prints the correct value
console.log("Supabase Service Role Key:", supabaseServiceRoleKey);  // Check if this prints the correct value

if (!supabaseUrl || !supabaseServiceRoleKey) {
  throw new Error("Supabase URL or Service Role Key not set in environment variables");
}

export const supabase = createClient(supabaseUrl, supabaseServiceRoleKey);


So all the above three files are placed under the lib directory.

Finally under the the page where you want to insert

const insertNewUser = async () => {
    const { data, error } = await supabaseAdmin.from('Users').insert({
      name: 'Sanjay',
      email: 's@example.com',
      address: 'UK',
      // Include other fields as required
    })
  
    if (data) {
      console.log('Supabase User data', data)
    }
  
    if (error) {
      console.log('Supabase Error', error)
    }
  }
  
 
  insertNewUser();

Create auth policies for your table.

Thanks

**Reply by TheLexoPlexx** [reply]
`NEXT_PUBLIC_SUPABASE_SERVICE_ROLE_KEY` is a genuine security risk, anon can be public, service-key should not be.

**Comment by Giorgi-Gamgebeli** [comment]
This doesn't resolve issue that key is unsafe even with row level security someone can still upload or access images

**Comment by muwew** [comment]
Hi! I had the same issue where everything was configured correctly but Vercel didn't seem to pick up the environment variables: turns out Vercel has a preview and production mode. In the environment variable dashboard in Vercel, I made sure the relevant environment variables can be run in all environments (tick all boxes including preview, production, development) and it worked for me!

**Comment by ShakurJuniior** [comment]
still facing the same issue, im using vite, my env is in the root folder,  the supabaseUrl still reading as undefined

**Comment by romuloTolfo** [comment]
prod funciona, development nao

**Comment by MrVibe** [comment]
its 2025 and I encountered this issue again with next 15.5. No clue what fixed it earlier when i faced it in 2023.

**Reply by yuxxeun** [reply]
can i look your code? this is my code using Next `15.3` and its fine so far, hopefully work for you too

<img width="460" height="137" alt="image" src="https://github.com/user-attachments/assets/97c4b82c-e87d-4a73-966b-4404c354bc59" />
